front: operational studies only display selected train conflicts

Caracol3 opened the pull request:
closes https://github.com/osrd-project/osrd-confidential/issues/1196

Uriel-Sautron commented on front/src/modules/conflict/styles/_conflictList.scss:
```diff
@@ -191,3 +175,42 @@
   display: none;
 }
+
+$train-category-colors: (
+  high-speed: 'red',
+  intercity: 'violet',
+  regional: 'orange',
+  night: 'purple',
+  commuter: 'blue',
+  freight: 'green',
+  fast-freight: 'mint',
+  tram-train: 'cyan',
+  touristic: 'brown',
+  work: 'yellow',
+  none: 'grey',
+);
```
This variable already exists in operationalStudies/_scenario.scss

Caracol3 replied:
You're right. Done

Uriel-Sautron commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -1,34 +1,86 @@
 }) => {
   const { t } = useTranslation('operational-studies', { keyPrefix: 'main' });
+  const dispatch = useDispatch();
+  const [visibleTrains, setVisibleTrains] = useState<number[]>([]);
+  const [showOthers, setShowOthers] = useState(false);
+  const [isOthersTooltipOpen, setIsOthersTooltipOpen] = useState(false);
+
+  const initialVisibleTrains = [...Array(conflict.trainNames.length).keys()];
```
```suggestion
  const [visibleTrains, setVisibleTrains] = useState<number[]>([...Array(conflict.trainNames.length).keys()]);
  const [showOthers, setShowOthers] = useState(false);
  const [isOthersTooltipOpen, setIsOthersTooltipOpen] = useState(false);
```

Caracol3 replied:
Good catch! Done

Uriel-Sautron commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -45,8 +97,13 @@ const ConflictCard = ({
+
+          if (!trainsToDisplay.includes(idx)) {
+            return null;
+          }
```
```suggestion
          if (!visibleTrains.includes(idx)) {
            return null;
          }
```

Caracol3 replied:
done

kmer2016 commented on front/src/modules/conflict/components/ConflictsList.tsx:
```diff
@@ -13,20 +14,132 @@ type ConflictsListProps = {
+    if (!selectedTrain && selectedTrainId.startsWith('indexedoccurrence_')) {
+      const pacedTrainId = `paced_${selectedTrainId.split('_')[1]}`;
+      selectedTrain = timetableItems.find((item) => item.id === pacedTrainId);
+    }
```
We can use existings [utils](https://github.com/OpenRailAssociation/osrd/blob/dev/front/src/utils/trainId.ts) function
```suggestion
    if (!selectedTrain && isIndexedOccurrenceId(selectedTrainId)) {
      const pacedTrainId = extractPacedTrainIdFromOccurrenceId(selectedTrainId);
      selectedTrain = timetableItems.find((item) => item.id === pacedTrainId);
    }
```

Caracol3 replied:
Done

kmer2016 commented on front/src/modules/conflict/components/ConflictsToolbar.tsx:
```diff
@@ -0,0 +1,61 @@
+      );
+    }
+
+    return (
+      <>
+        <span className="filter-text">pour </span>
+        <span className="train-name">{selectedTrainName}</span>
+        <span className="filter-text"> ({conflictsCount} conflits)</span>
+      </>
+    );
+  };
```
This bloc is the same that the one inside the if above

Caracol3 replied:
I added a ternary to handle the singular/plural case for 'conflict'.

kmer2016 commented on front/src/modules/conflict/components/ConflictsToolbar.tsx:
```diff
@@ -0,0 +1,61 @@
+          <span className="filter-text">pour </span>
+          <span className="train-name">{selectedTrainName}</span>
+          <span className="filter-text"> ({conflictsCount} conflit)</span>
+        </>
```
translating missing. `filter-text`class seems to be not used

Caracol3 replied:
Added new translations

kmer2016 commented on front/src/modules/conflict/components/ConflictsList.tsx:
```diff
@@ -13,20 +14,132 @@ type ConflictsListProps = {
+    return selectedTrain?.name || null;
+  }, [selectedTrainId, timetableItems]);
+
+  const displayedConflicts = useMemo(() => {
```
this function contains redundant code find in `useConflictsFilter`. I think we can refactor the hook and expose only pre-calculated data useful here

Caracol3 replied:
Done

kmer2016 commented on front/src/modules/conflict/components/ConflictsList.tsx:
```diff
@@ -13,20 +14,132 @@ type ConflictsListProps = {
+              (occurrenceNumber - baseNumber) % 2 === 0 && occurrenceNumber > baseNumber;
+            return isOccurrence;
+          }
+          return false;
```
can be move into util function

Caracol3 replied:
Done

kmer2016 commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -22,15 +27,15 @@ const useConflictsFilter = (
-    let selectedTrain = timetableItemsWithDetails.find((item) => item.id === selectedTrainId);
+    let selectedTrain = timetableItems.find((item) => item.id === selectedTrainId);
 
     if (!selectedTrain && selectedTrainId.startsWith('indexedoccurrence_')) {
```
you forgot to use util functions from `src/utils/trainId.ts `as `isIndexedOccurrenceId`, `formatEditoastIdToPacedTrainId`

Caracol3 replied:
Done

kmer2016 commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -41,41 +46,41 @@ const useConflictsFilter = (
-        }
-
-        return enrichedConflict.trainNames.includes(selectedTrainName);
+        const enrichedConflict = enrichSingleConflict(conflict, timetableItems);
```
In order to improve performance  by eliminating redundant computations, we can move `enrichedConflicts` to a separate useMemo.  it ensures conflicts are enriched only once and reused across both count calculation and display logic.

Caracol3 replied:
Done

kmer2016 commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -39,39 +36,36 @@ const useConflictsFilter = (
+    return result;
+  }, [enrichedConflicts, selectedTrainName, selectedTrainId]);
+
   const selectedTrainConflictsCount = useMemo(() => {
```
We can directly write `const selectedTrainConflictsCount = selectedEnrichedConflicts.length;` and remove useMemo, as it is only accessing an array property that is already memoised.

Caracol3 replied:
done

kmer2016 reviewed (approved):
small suggestion, but it's ok ! Great PR, Thank you ! Just don't forget to clean commit history

kmer2016 commented on front/src/applications/operationalStudies/views/Scenario/components/Timetable/Timetable.tsx:
```diff
@@ -141,6 +145,46 @@ const Timetable = ({
+    if (selectedTrainId && selectedTrainId !== previousSelectedTrainId.current) {
+      const isOccurrence =
+        selectedTrainId.startsWith('indexedoccurrence_') ||
+        selectedTrainId.startsWith('exception_');
```
We can use `isAddedExceptionId` and `isIndexedOccurrenceId` from traindId.ts
```suggestion
      const isOccurrence =
        isIndexedOccurrenceId(selectedTrainId) ||
        isAddedExceptionId(selectedTrainId)
```

Caracol3 replied:
Good catch ! Done

kmer2016 commented on front/src/modules/conflict/utils.ts:
```diff
@@ -95,3 +109,166 @@ export default function addTrainNamesToConflicts(
+): ConflictWithTrainNames {
+  const enrichedConflicts = addTrainNamesToConflicts([conflict], timetableItems);
+  return enrichedConflicts[0];
+}
```
can be drop since is not needed anymore

Caracol3 replied:
Done

kmer2016 commented on front/src/modules/conflict/utils.ts:
```diff
@@ -95,3 +109,166 @@ export default function addTrainNamesToConflicts(
+      }
+    }
+  }
+}
```
Since this function handles logic that is very specific to the ConflictCard component and is not reused elsewhere, it would be better to move it directly inside the component rather than exporting it from the utils.

Why we can use directly the id instead of the name of train ?

I noticed that we often need to look up timetable items by their ID, which currently requires iterating over the entire array each time. For better performance and code clarity, it would be more efficient to store timetable items in a Map (id → item) rather than a plain array, at least within functions that do multiple lookups.

For now, I suggest building the Map locally inside this function to avoid repeated array traversals. However, for a more robust and scalable solution, it would be even better to create the Map at a higher level and pass it down wherever timetable lookups are needed, instead of passing the array. This would improve performance throughout the codebase and make the intent clearer.

Caracol3 replied:
I moved the click logic into ConflictCard. We no longer rely on train names. The chip index maps directly to conflict IDs.

kmer2016 commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -61,11 +110,39 @@ const ConflictCard = ({
+            offsetRatio={{ top: 1.2 }}
+            reverseIfOverflow
+          />
+        )}
```
Factorized the "Others" button and tooltip logic using a React fragment (`<>` `</>)` to improve readability and show the logical relationship between these elements.

```suggestion
        {/* Show the "Others" card if there are more trains than visible */}
        {showOthers && (
          <>
            <div
              ref={othersTagRef}
              className="train-name-card train-category-text-None other-trains"
              onMouseEnter={() => setIsOthersTooltipOpen(true)}
              onMouseLeave={() => setIsOthersTooltipOpen(false)}
            >
              <span>
                {t('conflicts.otherTrains', {
                  count: conflict.trainNames.length - visibleTrains.length,
                })}
              </span>
            </div>
            
            {isOthersTooltipOpen && (
              <OSRDTooltip
                containerRef={othersTagRef}
                isOpen={true}
                header={t('conflicts.trainsInConflictTitle')}
                items={conflict.trainNames.filter((_, index) => !visibleTrains.includes(index))}
                offsetRatio={{ top: 1.2 }}
                reverseIfOverflow
              />
            )}
          </>
        )}
```

Caracol3 replied:
Done

kmer2016 commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -0,0 +1,75 @@
+
+    if (!selectedTrain && isIndexedOccurrenceId(selectedTrainId)) {
+      const pacedTrainId = extractPacedTrainIdFromOccurrenceId(selectedTrainId);
+      selectedTrain = timetableItems.find((item) => item.id === pacedTrainId);
```
Currently, we use Array.find on `timetableItems` each time we need to resolve the selected train. Since timetableItems can be large, this introduces an unnecessary lookup at every render.

We can build a Map<id, timetableItem> once (with useMemo) and use map.get(id) for lookups. This will  improve performance.

Caracol3 replied:
Done

kmer2016 commented on front/src/modules/conflict/utils.ts:
```diff
@@ -1,11 +1,21 @@
 import type { Conflict, TrainCategory } from 'common/api/osrdEditoastApi';
```
We can add some unit tests for these utils functions

Caracol3 replied:
Done

kmer2016 reviewed (commented):
Sorry another suggestions,

SharglutDev commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -0,0 +1,86 @@
+    [timetableItems]
+  );
+
+  const selectedTrainName = useMemo(() => {
```
This should rather be called `selectedTimetableItemName` since we return the train schedule name or paced train name

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -0,0 +1,86 @@
+    showOnlySelectedTrain,
+    handleToggleConflictsFilter,
+    selectedTrainName,
+    selectedTrainId,
```
This value is not used in `ScenarioContent`

Caracol3 replied:
Removed

SharglutDev commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -0,0 +1,86 @@
+import addTrainNamesToConflicts, { isTrainInConflict, reorderTrainsInConflict } from '../utils';
+
+const useConflictsFilter = (
+  timetableItems: TimetableItemWithDetails[],
```
nit: seems like we do not use the "details" part of `TimetableItemWithDetails`, we could only pass `TimetableItem[]` which is provided in `ScenarioContent`

Caracol3 replied:
Don't we need it when we use `addTrainNamesToConflicts` ?

SharglutDev replied:
We use `name`, `id` and `exceptions` which are all included in `TimetableItem` I think ?

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -0,0 +1,86 @@
+  );
+
+  const selectedEnrichedConflicts = useMemo(() => {
+    if (!selectedTrainName) return [] as ConflictWithTrainNames[];
```
Not needed
```suggestion
    if (!selectedTrainName) return [];
```

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/utils.ts:
```diff
@@ -95,3 +101,72 @@ export default function addTrainNamesToConflicts(
+): boolean {
+  if (!selectedTrainId) return false;
+
+  if (isPacedTrainId(selectedTrainId) || isIndexedOccurrenceId(selectedTrainId)) {
```
We can only select a train schedule or an occurrence so `isPacedTrainId(selectedTrainId)` is always `false` for a selected train (same for `reorderTrainsInConflict` function)

Caracol3 replied:
You're right ! Done

SharglutDev commented on front/src/modules/conflict/utils.ts:
```diff
@@ -95,3 +101,72 @@ export default function addTrainNamesToConflicts(
+  if (Number.isNaN(Number(selectedTrainName)) || Number.isNaN(Number(trainName))) {
+    return false;
+  }
+  return (occurrenceNumber - baseNumber) % 2 === 0 && occurrenceNumber > baseNumber;
```
What if the `selectedTrainName` is 7618 and `trainName` (one of the conflicted occurrence) is 8602 ?
This will return `true` but I don't think it should be a match ?

If I understand correctly, when we check the selected train name (name of `TrainId`) checkbox, we want to filter among all conflicts only the ones where the selected train name is involved.

In `addTrainNamesToConflicts`, we add in `trainNames` property either the train schedule names or the occurrence names (for paced train).

`selectedTrainName` in `useConflictsFilter` compute the paced train name if the selected train id is an occurrence but the issue mentions `filtrer la liste des conflits pour ne voir que ceux du train sélectionné`. The selected train is not a paced train but an occurrence or did I misunderstand the issue ? (hi @maelysLeratRosso)

If we mean the occurrence name then we can only check for `conflict.trainNames.includes(selectedTrainName)`.

If we in fact mean the paced train name (and thus want to filter all conflicts with occurrences belonging to the paced train corresponding to the selected occurrence), we probably gonna need to find a way to connect each of the `conflict.trainNames`occurrences with their corresponding paced train, which can probably only be the `pacedTrainId`.

Caracol3 replied:
I removed `isTrainMatch`which indeed wasn't perfect. 
For `isTrainInConflict`: switched from name-based parity matching to stable ID checks (occurrence → paced_train_id, schedule → numeric id) to avoid false positives like ‘8602’ vs ‘7618’.
For `reorderTrainsInConflict`: only reorders labels for display:  for occurrences we first try exact base-name, then occurences prefix base. Schedules rely on exact name.

SharglutDev commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -0,0 +1,86 @@
+  );
+
+  const selectedEnrichedConflicts = useMemo(() => {
+    if (!selectedTrainName) return [] as ConflictWithTrainNames[];
```
We could also add `|| !selectedTrainId` and then remove the `undefined` case of `selectedTrainId` param for `isTrainInConflict` and `reorderTrainsInConflict` functions.

Caracol3 replied:
Good idea !
Done

SharglutDev commented on front/src/modules/conflict/utils.ts:
```diff
@@ -95,3 +101,72 @@ export default function addTrainNamesToConflicts(
+    // Remove the selected train from its current position
+    trainNames.splice(selectedTrainIndex, 1);
+    trainCategories.splice(selectedTrainIndex, 1);
+
+    // Add it to the front
+    trainNames.unshift(trainNameToMove);
+    trainCategories.unshift(trainCategoryToMove);
+  }
```
Instead of countingon the fact that each conflicted train is at the same index in `trainNames` and `trainCategories`, would it be safier if the type was more like : 
```
type ConflictWithTrainNames = Conflict & {
  trainsData: {
     name: string;
     category: TrainCategory | null
  }[]
};
```

Caracol3 replied:
Safer now. Refactored to a single trainsData array to keep name and category together

SharglutDev commented on front/src/modules/conflict/components/Conflicts.tsx:
```diff
@@ -0,0 +1,44 @@
+import type { ConflictWithTrainNames } from '../types';
+
+type ConflictsProps = {
+  conflicts: Conflict[];
```
Not used
```suggestion
```

Caracol3 replied:
Removed

SharglutDev commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -0,0 +1,86 @@
+
+  const selectedTrainConflictsCount = selectedEnrichedConflicts.length;
+
+  const getDisplayedConflicts = useCallback(() => {
```
Can't we just return `displayedConflicts` / `filteredConflicts` instead of `getDisplayedConflicts`  function that we gonna need to resolve in `ConflictsList` ?

Caracol3 replied:
Good catch !
Done

SharglutDev commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -19,16 +24,69 @@ const ConflictCard = ({
 
+  const trainIds = useMemo(() => {
+    const scheduleIds = conflict.train_schedule_ids.map(
+      (id) => `trainschedule_${id}` as TrainScheduleId
```
Same for occurrences below
```suggestion
      (id) => formatEditoastIdToTrainScheduleId(id)
```

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -19,16 +24,69 @@ const ConflictCard = ({
+    const id = trainIds[trainIndex];
+    if (!id) return;
+    dispatch(updateSelectedTrainId(id));
+  };
```
How does it work if we are in "selected train conflicts filter mode" ? We click on one of the train name in the conflict card, it changes the selected train id so it totally changes the UI of the conflict list since the checkbox is checked ?

Caracol3 replied:
Selection from conflicts is now removed

SharglutDev commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -19,16 +24,69 @@ const ConflictCard = ({
+
+  const onTrainClick = (trainIndex: number, event: React.MouseEvent) => {
+    event.stopPropagation();
+    const id = trainIds[trainIndex];
```
How do you know that the proper train will be selected ? It does not seem like `trainIds` are ordered the same way as `conflict.trainNames`.

Caracol3 replied:
Selection from conflicts is now removed

SharglutDev commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -19,16 +24,69 @@ const ConflictCard = ({
+    const visibleTrainIndices: number[] = [];
+
+    trainCardElements.forEach((cardElement, trainIndex) => {
+      const trainCardElement = cardElement as HTMLElement;
```
Not needed I think
```suggestion
```

Caracol3 replied:
Removed

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/Timetable/Timetable.tsx:
```diff
@@ -141,6 +147,45 @@ const Timetable = ({
+        });
+      }
+    }
+  }, [selectedTrainId]);
```
If we select a train directly in the timetable which is not in the center of the screen, won't it activate this effect ? (not tested)

Caracol3 replied:
Selection from conflicts is now removed

SharglutDev reviewed (commented):
Great work overall, this enhancement is quite a fat one !

Sorry for the late review but was checking the code and had some comments to make. Don't forget to change the code in the concerned commits.

SharglutDev commented on front/src/modules/conflict/utils.ts:
```diff
@@ -95,78 +97,64 @@ export default function addTrainNamesToConflicts(
-    return conflict.trainNames.some((trainName) => isTrainMatch(trainName, selectedTrainName));
+  const selectedId = extractEditoastIdFromTrainId(selectedTrainId);
+
+  if (isOccurrenceId(selectedTrainId) || isIndexedOccurrenceId(selectedTrainId)) {
```
`isOccurrenceId` tests for both indexed occurrences and added occurrences

```suggestion
  if (isOccurrenceId(selectedTrainId)) {
```

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/utils.ts:
```diff
@@ -95,78 +97,64 @@ export default function addTrainNamesToConflicts(
+  const selectedId = extractEditoastIdFromTrainId(selectedTrainId);
+
+  if (isOccurrenceId(selectedTrainId) || isIndexedOccurrenceId(selectedTrainId)) {
+    return conflict.paced_train_occurrence_ids.some((occ) => occ.paced_train_id === selectedId);
```
With this change, we only check for a paced train match but wouldn't it be enough to check the occurrence ?

Each enriched conflict has all the train names involved in `conflict.trainsData`. We could loop on this to find a match with the selected train name and it would be enough to filter it, no matter if it's a train schedule or an occurrence.

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/utils.ts:
```diff
@@ -95,78 +97,64 @@ export default function addTrainNamesToConflicts(
+  return false;
 }
 
 export function reorderTrainsInConflict(
```
If I understand correctly, we always to put the selected train name at first of the list in the train name list involved in the conflict.
But we already check for that in `isTrainInConflict`.

To simplify the code, we could merge `isTrainInConflict` and `reorderTrainsInConflict` into one function which would take `selectedTrainName` and `enrichedConflict` and would return either `null` if the conflict doesn't included the selected train name or the conflict itself with its train names already sorted with the selected one at first.

What do you think ?

Caracol3 replied:
`isTrainInConflict()` and `reorderTrainsInConflict()` are now merged into a single `filterAndReorderConflict()` function that handles both filtering and reordering in one step.

SharglutDev commented on front/src/modules/conflict/components/ConflictsList.tsx:
```diff
@@ -1,23 +1,17 @@
-    </div>
-  );
-};
+const ConflictsList = ({ onConflictClick, displayedConflicts }: ConflictsListProps) => (
```
Since this component doesn't have any logic left in it maybe we could get rid of it and call its content directly in `Conflicts.tsx` ?

SharglutDev reviewed (commented):
For the remove of the selection logic myabe you could remove the 2 related new commits in the PR and then get rid off the old related code.

SharglutDev commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -1,34 +1,64 @@
 
   const subCategories = useSubCategoryContext();
 
+  useLayoutEffect(() => {
```
There is a weird behavior sometimes. There is a lot of space to display more conflict but we see the "+x" tag instead.

<img width="399" height="514" alt="Image" src="https://github.com/user-attachments/assets/d2c14e72-6974-480c-8943-a7ad1bf6d600" />

Caracol3 replied:
Is this a timetable you still have at hand ?

SharglutDev replied:
It was the stdcm scenario

SharglutDev replied:
But my laptop was struggling with all the computing maybe some weird stuff happened. Curious if you can reproduce.

Caracol3 replied:
I will try with the STDCM scenario. I personally have never experienced this bug.

Caracol3 replied:
No problem on my side.

<img width="382" height="489" alt="image" src="https://github.com/user-attachments/assets/5676aa80-a09a-47f4-b14b-580912e1e1f3" />

SharglutDev replied:
I still have it 
<img width="389" height="884" alt="Capture d’écran 2025-10-02 à 16 30 30" src="https://github.com/user-attachments/assets/99991c81-c4d9-4cd8-82de-e353199bb969" />

Here is the timetable I used
[new_conflicts.json](https://github.com/user-attachments/files/22661522/new_conflicts.json)

Something might be wrong in the `useLayoutEffect` because there is enough space to display the extra train name.

Caracol3 replied:
We now calculate layout ourselves instead of letting the browser handle it, because Firefox wasn't handling `getBoundingClientRect()` + `clip-path` well.

SharglutDev commented on front/src/modules/conflict/styles/_conflicts.scss:
```diff
@@ -132,27 +150,10 @@
-        display: block;
+        &:hover::after {
+          background-color: var(--grey10);
+        }
```
Since we removed the selection behavior, we can probably remove this style as well

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -0,0 +1,107 @@
+    [timetableItems]
+  );
+
+  const selectedTimetableItemName = useMemo(() => {
```
```suggestion
  const selectedTrainName = useMemo(() => {
```

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -0,0 +1,107 @@
+      return selectedTrain?.train_name || null;
+    }
+
+    if (isOccurrenceId(selectedTrainId)) {
```
We don't need this condition. `selectedTrainId` is always an `OccurrenceId` at this point (we can also remove the last `null` return).

Caracol3 replied:
You're right. Removed

SharglutDev commented on front/src/modules/conflict/utils.ts:
```diff
@@ -71,27 +72,70 @@ function getConflictTrainCategories(
+    const selectedId = extractEditoastIdFromTrainId(selectedTrainId);
+    const isTrainMatch = conflict.train_schedule_ids.includes(selectedId);
+    if (!isTrainMatch) return null;
+  }
```
The first check above is not enough for train schedules ?

Caracol3 replied:
Indeed, I removed it. I also changed the first check to strict equality on the train name instead of `includes()` to avoid substring false positives and ensure exact matches.

SharglutDev replied:
I still see the `if (!isOccurrenceId(selectedTrainId))` condition

SharglutDev commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -0,0 +1,107 @@
+      .map((conflict) =>
+        filterAndReorderConflict(conflict, selectedTrainId, selectedTimetableItemName)
+      )
+      .filter((conflict): conflict is ConflictWithTrainNames => conflict !== null);
```
```suggestion
      .filter((conflict) => conflict !== null);
```

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/hooks/useConflictsFilter.ts:
```diff
@@ -0,0 +1,107 @@
+  const selectedEnrichedConflicts = useMemo(() => {
+    if (!selectedTimetableItemName || !selectedTrainId) return [];
+    return enrichedConflicts
+      .map((conflict) =>
+        filterAndReorderConflict(conflict, selectedTrainId, selectedTimetableItemName)
+      )
+      .filter((conflict): conflict is ConflictWithTrainNames => conflict !== null);
```
My `reduce` clock is ticking but I won't gonna say anything 😜

SharglutDev commented on front/src/modules/conflict/__tests__/utils.spec.ts:
```diff
@@ -0,0 +1,70 @@
+
+describe('addTrainNamesToConflicts', () => {
+  it('combines schedule and occurrence names', () => {
+    const trains: TimetableItemWithDetails[] = [
```
We need to change the type here since we now pass `TimetableItem` (and update the other tests after the change of the util functions)

Caracol3 replied:
Done

Caracol3 committed:
front: allow to filter on selected train conflicts

- show others tag if there are more train than the ones visible
- add tooltip when hovering others tag in conflictlist
- when filtered display selected train first on conflict card

Signed-off-by: Mathieu <mathieu.coulibaly@sncf.fr>

kmer2016 commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -1,34 +1,64 @@
+
+      if (isTrainCardFullyVisible) {
+        visibleTrainIndices.push(trainIndex);
+      }
```
nit
```suggestion
      if (isHorizontallyVisible && isVerticallyVisible) {
        visibleTrainIndices.push(trainIndex);
      }
```

Caracol3 replied:
Done

kmer2016 reviewed (approved):
lgtm

SharglutDev commented on front/public/locales/fr/operational-studies.json:
```diff
@@ -135,12 +139,13 @@
     "pacedTrainAndTrainCount": "$t(main.pacedTrainCount, { 'count': {{pacedTrainCount}}, 'totalCount': {{totalPacedTrainCount}} }), $t(main.trainCount, { 'count': {{trainCount}}, 'totalCount': {{totalTrainScheduleCount}} }) sélectionnés",
     "pacedTrainCountSelected_one": "1/{{totalCount}} mission sélectionnée",
     "pacedTrainCountSelected_other": "{{count}}/{{totalCount}} missions sélectionnées",
+    "pacedTrainCountSelected_zero": "$t(main.pacedTrain, { 'count': {{totalCount}} })",
```
Where are these changes coming from ?

Caracol3 replied:
The i18n CI checker was failing due to this missing translation key. Added it to fix the build pipeline.

SharglutDev replied:
Yeah but you didn't make any changes related to these keys, why does the CI fails now ? If you remove them, the CI still fails ?

Caracol3 replied:
The CI indeed doesn't seem to fail now that I removed this line. I added this one among others, but it apparently wasn’t necessary.

SharglutDev replied:
I think we can remove also `trainCountSelected_zero` ?

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -1,34 +1,62 @@
+
+    trainCardElements.forEach((cardElement, trainIndex) => {
+      const trainCardBounds = cardElement.getBoundingClientRect();
+      const containerBounds = container.getBoundingClientRect();
```
Maybe we should move this line outside of the loop. It does not change across time, doesn't it ?

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -1,34 +1,90 @@
+
+    const calculateVisibleTrains = () => {
+      const container = trainsContainerRef.current;
+      if (!container) return;
```
We can maybe remove the same test above the function if we need it inside it again anyway

Caracol3 replied:
Done

SharglutDev commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -1,34 +1,90 @@
+        const cardWidth = cardElement.offsetWidth;
+
+        const wouldExceedWidth =
+          currentRowWidth + cardWidth + (currentRowWidth > 0 ? GAP : 0) > containerWidth;
```
Almost perfect ! In small screens, it seems the `+x others` is just not displayed. It exists in the DOM though but it is overflowing.

To see it, you can take the timetable I shared before and check the conflict with `start exception` and `coucou label` (don't mind the names 🙄)

SharglutDev replied:
Not sure I like the new logic better. Since it is an edge case, I would vote to keep the state like it was before your last commit, merge like that and make an enhancement later so you could move on from this issue.

SharglutDev commented on front/src/modules/conflict/components/ConflictCard.tsx:
```diff
@@ -44,9 +46,12 @@ const ConflictCard = ({
-          const category = conflict.trainCategories[idx];
+      <div className="trains-name" ref={trainsContainerRef}>
+        {conflict.trainsData.map((train, idx) => {
+          if (!visibleTrains.includes(idx)) {
+            return null;
+          }
```
We could also filter the trains with `idx > 2` and not use `visibleTrains` I think ? And `idx < 3` for the tooltip.

Caracol3 replied:
Using `idx > 2` hides the 4th train when there are exactly 4 trains (no “Others”).
Now we show all tags when `total ≤ 4`. When `total > 4`, render the first 3 and “Others”, with the tooltip listing trains where `idx ≥ 3`.

SharglutDev replied:
Yeah I meant we could use it combined with `showOthers` value. But it is not that different than your approach, we can leave it like that if you prefer.

Caracol3 replied:
I'm fine with the two options so leave it like that is good for me.

SharglutDev reviewed (approved):
Lgtm ✅ Really nice work ! 🙂
